### ISSUE TITLE
Questions for the paper.

### ISSUE THREAD

**Issue by minhquoc0712** [paulgavrikov/CNN-Filter-DB#3]
Thank you for your interesting research.

I would be very appreciate if you can answer the following questions so that I can better understand your work.

1. Can you release the code for Figure 1 in your paper?
2. Should the first term in equation (3) have sigma_i? And what is the difference between f_i' and f_i?
3. If I understand correctly from your code, you use `conv_depth_norm` as overparameterization in Figure 2. I don't understand why a higher `conv_depth_norm` would represent overparameterization. Since if you have a shallow network, the last layers still have large `conv_depth_norm`. I may miss something here.
4. How do you choose the sparsity threshold in Eq. (6)? Should it be changed across layers and models?
5. For the bi-variate plot in Figure 4, what are the components for each plot? It is unclear to me how you come up with your comment from the shape of each plot, especially the second and third ones. Can you give me a clearer explanation?

Thank you.

**Comment by paulgavrikov** [comment]
Thank you for your interest in our work! 

1. Here is the code we used to generate the plot. Adjust it to your model as needed.

```python
import matplotlib.pyplot as plt
from matplotlib import lines
from matplotlib.colors import LinearSegmentedColormap
import torch

def plot_filters_from_series(checkpoint, cols=28, figsize=(8, 5)):
    rows = len(checkpoint)
    fig, axes = plt.subplots(rows, cols, figsize=figsize, squeeze=False)
    for row_idx, (row, layer) in enumerate(zip(axes, checkpoint)):
        layer = layer / abs(layer).max()        
        for i, (ax, f) in enumerate(zip(row, layer)):
            ax.spines['bottom'].set_visible(False)
            ax.spines['top'].set_visible(False) 
            ax.spines['right'].set_visible(False)
            ax.spines['left'].set_visible(False)
            ax.xaxis.set_major_locator(plt.NullLocator())
            ax.yaxis.set_major_locator(plt.NullLocator())
            ax.imshow(f.reshape(3, 3), vmin=-0.3, vmax=0.3, cmap=LinearSegmentedColormap.from_list("CyanOrange", ["C0", "white", "C1"]))
            
def get_filters(name, path):
    # TODO: create the model and load the state dict
    
    filters = []
    for i, m in enumerate(filter(lambda m: type(m) == torch.nn.Conv2d and m.kernel_size == (3, 3), model.modules())):
        w = m.weight.detach().cpu().numpy().reshape(-1, 9).copy()
        filters.append(w)
    return filters
    
    
plot_filters_from_series(get_filters(...))
```

2. Yes, I think c_i = sigma_i. I don't recall why we used a new symbol here. Regarding f': For this equation, you fully reconstruct the initial input - so f' is X_i. I think we initially had a different order of the methods so that seems like a leftover notation and you can just think of it as f. For the analysis of patterns (filter structure) in Sec. 4.2 we used f' to denote the normalized filters.

3. `conv_depth_norm` is the relative depth of the layer in the model. We do not use it as a metric (it's the x-axis in Fig. 2). We use the sparsity and entropy metrics from Sec 3.4 as y-axis. By now, we released a module on pypi to calculate it (as well as some other metrics): https://github.com/paulgavrikov/torchconvquality

4. We have a discussion about that in Sec 4.1 ("Thresholds") and the Supplementary Materials Sec C. But the short answer is: there is no right or wrong here - we empirically found it. E.g. for the ablation is the Supp we used 0.14. I don't think it makes sense to pick the threshold for every layer since it is always relative to the layer max. Ideally, you'd apply magnitude pruning to the model and find this value at the point where the accuracy starts to fall off but that was too expensive for our dataset.

5. For space reasons we only used two random components to compute these plots: Ideally, you'd show a plot between all combinations of components (see Supp. Fig.12ff). Effectively, these are just 2D distributions of the distributions shown in Fig. 3. If your model has a "healthy" variety in filter patterns then you will see a "sun" phenotype - that is both components are approx. gaussian distributed and the plot will be a smooth circle/oval. If your model degenerates it will only learn specific filter patterns visible by spikes in the KDEs: you will start to see local hotspots ("spikes"/"point")  or a distribution that does not fully exploit the space ("symbols").


Hope that helps! Let me know if you have any other questions!
